state/sqlserver/migration.go

```diff
@@ -275,32 +274,61 @@ func (m *migration) ensureUpsertStoredProcedureExists(db *sql.DB, mr migrationRe
 		CREATE PROCEDURE %s (
 			@Key 			%s,
 			@Data 			NVARCHAR(MAX),
-			@RowVersion 	BINARY(8))
+			@RowVersion 	BINARY(8),
+			@FirstWrite		BIT)
 		AS
 			IF (@RowVersion IS NOT NULL)
 			BEGIN
 				UPDATE [%s]
 				SET [Data]=@Data, UpdateDate=GETDATE()
 				WHERE [Key]=@Key AND RowVersion = @RowVersion
-
 				RETURN
 			END
 
-			BEGIN TRY
-				INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
-			END TRY
-
-			BEGIN CATCH
-				IF ERROR_NUMBER() IN (2601, 2627)
-				UPDATE [%s]
-				SET [Data]=@Data, UpdateDate=GETDATE()
-				WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
-			END CATCH`,
+			IF (@FirstWrite=1)
+			BEGIN
+				BEGIN TRANSACTION;
+				IF EXISTS (SELECT * FROM [%s] WHERE [KEY]=@KEY)
+				BEGIN
+					THROW 2714, ''FIRST-WRITE: COMPETING RECORD ALREADY WRITTEN.'', 1
+					RETURN
+				END
+				BEGIN
+					BEGIN TRY
+						INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
+					END TRY
+		
+					BEGIN CATCH
+						IF ERROR_NUMBER() IN (2601, 2627)
+						UPDATE [%s]
+						SET [Data]=@Data, UpdateDate=GETDATE()
+						WHERE [Key]=@Key AND RowVersion = ISNULL(@RowVersion, RowVersion)
+					END CATCH
+				END
+				COMMIT;
+			END
+			ELSE
+			BEGIN
+				BEGIN TRY
+					INSERT INTO [%s] ([Key], [Data]) VALUES (@Key, @Data);
```

Comment:
In this case, FirstWrite=0 but RowVersion can be NULL or not, so it must be handled here too.

Reply:
The problem is that the stored procedure and DB has for 2 years always set a Rowversion / Etag. So Rowversion cannot actually be null in the DB, only in the request.

What I observed happening with the existing stored procedure was that any attempt to update a Key without also specifying the correct Rowversion failed (was ignored as NO-OP), unless the Rowversion was NULL. When the existing ROWVERSION was provided the record was updated and the ROWVERSION was incremeneted.

Reply:
Ignore my comment - I was confused about the First Write behavior. Will update.

Reply:
Please take a look at my latest change on this. I think I am handling it correctly now -- and conformance tests and integration tests pass.